File: packages/go/luna/resources/projects.go
package resources

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"regexp"
	"strconv"

	lunahttp "github.com/eclipse-softworks/luna-sdk-go/luna/http"
)

var projectIDPattern = regexp.MustCompile(`^prj_[a-zA-Z0-9]+$`)

// ProjectsResource provides access to project operations
type ProjectsResource struct {
	client   *lunahttp.Client
	basePath string
}

// NewProjectsResource creates a new projects resource
func NewProjectsResource(client *lunahttp.Client) *ProjectsResource {
	return &ProjectsResource{
		client:   client,
		basePath: "/v1/projects",
	}
}

// List retrieves all projects with pagination
func (r *ProjectsResource) List(ctx context.Context, params *ListParams) (*ProjectList, error) {
	query := url.Values{}
	if params != nil {
		if params.Limit > 0 {
			query.Set("limit", strconv.Itoa(params.Limit))
		}
		if params.Cursor != "" {
			query.Set("cursor", params.Cursor)
		}
	}

	resp, err := r.client.Request(ctx, lunahttp.RequestConfig{
		Method: "GET",
		Path:   r.basePath,
		Query:  query,
	})
	if err != nil {
		return nil, err
	}

	var result ProjectList
	if err := json.Unmarshal(resp.Data, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &result, nil
}

// Get retrieves a project by ID
func (r *ProjectsResource) Get(ctx context.Context, projectID string) (*Project, error) {
	if err := validateProjectID(projectID); err != nil {
		return nil, err
	}

	resp, err := r.client.Request(ctx, lunahttp.RequestConfig{
		Method: "GET",
		Path:   fmt.Sprintf("%s/%s", r.basePath, projectID),
	})
	if err != nil {
		return nil, err
	}

	var result Project
	if err := json.Unmarshal(resp.Data, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &result, nil
}

// Create creates a new project
func (r *ProjectsResource) Create(ctx context.Context, data ProjectCreate) (*Project, error) {
	if data.Name == "" {
		return nil, fmt.Errorf("name is required")
	}

	resp, err := r.client.Request(ctx, lunahttp.RequestConfig{
		Method: "POST",
		Path:   r.basePath,
		Body:   data,
	})
	if err != nil {
		return nil, err
	}

	var result Project
	if err := json.Unmarshal(resp.Data, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &result, nil
}

// Update updates an existing project
func (r *ProjectsResource) Update(ctx context.Context, projectID string, data ProjectUpdate) (*Project, error) {
	if err := validateProjectID(projectID); err != nil {
		return nil, err
	}

	resp, err := r.client.Request(ctx, lunahttp.RequestConfig{
		Method: "PATCH",
		Path:   fmt.Sprintf("%s/%s", r.basePath, projectID),
		Body:   data,
	})
	if err != nil {
		return nil, err
	}

	var result Project
	if err := json.Unmarshal(resp.Data, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &result, nil
}

// Delete deletes a project
func (r *ProjectsResource) Delete(ctx context.Context, projectID string) error {
	if err := validateProjectID(projectID); err != nil {
		return err
	}

	_, err := r.client.Request(ctx, lunahttp.RequestConfig{
		Method: "DELETE",
		Path:   fmt.Sprintf("%s/%s", r.basePath, projectID),
	})
	return err
}

func validateProjectID(id string) error {
	if id == "" {
		return fmt.Errorf("project ID is required")
	}
	if !projectIDPattern.MatchString(id) {
		return fmt.Errorf("invalid project ID format, expected: prj_<id>")
	}
	return nil
}

File: packages/typescript/src/resources/projects.ts
import type { HttpClient } from '../http/client.js';
import type {
    Project,
    ProjectCreate,
    ProjectUpdate,
    ProjectList,
    PaginationParams,
} from '../types/index.js';

/**
 * Projects resource for managing projects
 *
 * @example
 * ```typescript
 * // List projects
 * const projects = await client.projects.list({ limit: 10 });
 *
 * // Get a project
 * const project = await client.projects.get('prj_123');
 *
 * // Create a project
 * const newProject = await client.projects.create({
 *   name: 'My Project',
 *   description: 'A new project',
 * });
 *
 * // Update a project
 * const updated = await client.projects.update('prj_123', { name: 'Updated' });
 *
 * // Delete a project
 * await client.projects.delete('prj_123');
 * ```
 */
export class ProjectsResource {
    private readonly httpClient: HttpClient;
    private readonly basePath = '/v1/projects';

    constructor(httpClient: HttpClient) {
        this.httpClient = httpClient;
    }

    /**
     * List all projects with pagination
     */
    async list(params?: PaginationParams): Promise<ProjectList> {
        const response = await this.httpClient.request<ProjectList>({
            method: 'GET',
            path: this.basePath,
            query: {
                limit: params?.limit?.toString(),
                cursor: params?.cursor,
            },
        });

        return response.data;
    }

    /**
     * Get a project by ID
     */
    async get(projectId: string): Promise<Project> {
        this.validateProjectId(projectId);

        const response = await this.httpClient.request<Project>({
            method: 'GET',
            path: `${this.basePath}/${projectId}`,
        });

        return response.data;
    }

    /**
     * Create a new project
     */
    async create(data: ProjectCreate): Promise<Project> {
        this.validateProjectCreate(data);

        const response = await this.httpClient.request<Project>({
            method: 'POST',
            path: this.basePath,
            body: data,
        });

        return response.data;
    }

    /**
     * Update an existing project
     */
    async update(projectId: string, data: ProjectUpdate): Promise<Project> {
        this.validateProjectId(projectId);

        const response = await this.httpClient.request<Project>({
            method: 'PATCH',
            path: `${this.basePath}/${projectId}`,
            body: data,
        });

        return response.data;
    }

    /**
     * Delete a project
     */
    async delete(projectId: string): Promise<void> {
        this.validateProjectId(projectId);

        await this.httpClient.request<void>({
            method: 'DELETE',
            path: `${this.basePath}/${projectId}`,
        });
    }

    /**
     * Validate project ID format
     */
    private validateProjectId(projectId: string): void {
        if (!projectId) {
            throw new Error('Project ID is required');
        }
        if (!/^prj_[a-zA-Z0-9]+$/.test(projectId)) {
            throw new Error('Invalid project ID format. Expected: prj_<id>');
        }
    }

    /**
     * Validate project creation data
     */
    private validateProjectCreate(data: ProjectCreate): void {
        if (!data.name) {
            throw new Error('Name is required');
        }
    }
}
